openedx/features/content_tagging/rules.py

```diff
@@ -35,27 +36,108 @@ def is_taxonomy_admin(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool
+    Otherwise, a taxonomy must be provided to determine if the user is a org-level course creator for one of the
+    taxonomy's org owners.
+    """
+    if is_global_taxonomy_admin(user):
+        return True
+
+    if is_global_taxonomy(taxonomy):
+        return False  # Only global admins can edit global taxonomies
+
+    taxonomy_orgs = TaxonomyOrg.get_admin_organizations(
+        taxonomy=taxonomy,
+        rel_type=TaxonomyOrg.RelType.OWNER,
+    )
+    for org in taxonomy_orgs:
+        if is_content_creator(user, org.short_name):
+            return True
+    return False
+
+
+def is_global_taxonomy(taxonomy: oel_tagging.Taxonomy) -> bool:
+    """
+    Returns True if the given taxonomy is a global taxonomy.
+    """
+    return TaxonomyOrg.objects.filter(
+        taxonomy=taxonomy,
+        rel_type=TaxonomyOrg.RelType.OWNER,
+        org=None,
+    ).exists()
+
+
+def get_user_taxonomy_orgs(user: User) -> List[Organization]:
+    """
+    Return the list of organizations for which the given user is a taxonomy admin.
+    """
+    orgs = Organization.objects.all()
+
+    if is_global_taxonomy_admin(user):
+        return list(orgs)
+
+    return [org for org in orgs if is_content_creator(user, org.short_name)]
+
+
 @rules.predicate
 def can_view_taxonomy(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool:
     """
-    Anyone can view an enabled taxonomy,
-    but only taxonomy admins can view a disabled taxonomy.
+    Everyone can potentially view a taxonomy (taxonomy=None). The object permission must be checked
+    to determine if the user can view a specific taxonomy.
+    Only taxonomy admins can view a disabled taxonomy. #ToDo: Check if this is correct.
     """
-    if taxonomy:
-        taxonomy = taxonomy.cast()
-    return (taxonomy and taxonomy.enabled) or is_taxonomy_admin(user, taxonomy)
+    if not taxonomy:
+        return True
+
+    taxonomy = taxonomy.cast()
+
+    return is_taxonomy_admin(user, taxonomy) or taxonomy.enabled
+
+
+@rules.predicate
+def can_add_taxonomy(user: User) -> bool:
+    """
+    Only global taxonomy admins add taxonomies.
+    There is no API method to create a taxonomy associated with an organization
+    neither to add a taxonomy to an organization.
+    """
+    return is_global_taxonomy_admin(user)
 
 
 @rules.predicate
 def can_change_taxonomy(user: User, taxonomy: oel_tagging.Taxonomy = None) -> bool:
     """
+    Everyone can potentially change a taxonomy (taxonomy=None). The object permission must be checked
+    to determine if the user can change a specific taxonomy.
     Even taxonomy admins cannot change system taxonomies.
     """
-    if taxonomy:
-        taxonomy = taxonomy.cast()
-    return is_taxonomy_admin(user, taxonomy) and (
-        not taxonomy or (taxonomy and not taxonomy.system_defined)
-    )
+    if not taxonomy:
+        return True
+
+    taxonomy = taxonomy.cast()
+
+    return is_taxonomy_admin(user, taxonomy) and (not taxonomy.system_defined)
```

Comment:
Similar performance nit:
```suggestion
    return (not taxonomy.system_defined) and is_taxonomy_admin(user, taxonomy)
```

Reply:
Also, your change doesn't actually change the logic of this method. It does clarify it though, so I'm ok with keeping it. But in future, could you please avoid making unnecessary changes to keep the diff as small as possible?

Reply:
Sure! I remove some logic redundancies overall in the checks, but I agree with you. I will try to keep the changes small next time!